pkg/server/list_users.go

```diff
@@ -0,0 +1,131 @@
+	}
+
+	if !validator.RequestIsValidatedFromContext(ctx) {
+		if err := req.Validate(); err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+	}
+
+	const methodName = "listusers"
+
+	typesys, err := s.resolveTypesystem(ctx, req.GetStoreId(), req.GetAuthorizationModelId())
+	if err != nil {
+		telemetry.TraceError(span, err)
+		return nil, err
+	}
+
+	err = listusers.ValidateListUsersRequest(ctx, req, typesys)
+	if err != nil {
+		telemetry.TraceError(span, err)
+		return nil, err
+	}
+
+	ctx = typesystem.ContextWithTypesystem(ctx, typesys)
+
+	listUsersQuery := listusers.NewListUsersQuery(s.datastore,
+		listusers.WithResolveNodeLimit(s.resolveNodeLimit),
+		listusers.WithResolveNodeBreadthLimit(s.resolveNodeBreadthLimit),
+		listusers.WithListUsersQueryLogger(s.logger),
+		listusers.WithListUsersMaxResults(s.listUsersMaxResults),
+		listusers.WithListUsersDeadline(s.listUsersDeadline),
+		listusers.WithListUsersMaxConcurrentReads(s.maxConcurrentReadsForListUsers),
+	)
+
+	resp, err := listUsersQuery.ListUsers(ctx, req)
+	if err != nil {
+		telemetry.TraceError(span, err)
+
+		switch {
+		case errors.Is(err, graph.ErrResolutionDepthExceeded):
+			return nil, serverErrors.AuthorizationModelResolutionTooComplex
+		case errors.Is(err, condition.ErrEvaluationFailed):
+			return nil, serverErrors.ValidationError(err)
+		default:
+			return nil, serverErrors.HandleError("", err)
+		}
+	}
+
+	datastoreQueryCount := float64(resp.Metadata.DatastoreQueryCount)
+
+	grpc_ctxtags.Extract(ctx).Set(datastoreQueryCountHistogramName, datastoreQueryCount)
+	span.SetAttributes(attribute.Float64(datastoreQueryCountHistogramName, datastoreQueryCount))
+	datastoreQueryCountHistogram.WithLabelValues(
+		s.serviceName,
+		"list-users",
+	).Observe(datastoreQueryCount)
+
+	dispatchCount := float64(resp.Metadata.DispatchCounter.Load())
+	grpc_ctxtags.Extract(ctx).Set(dispatchCountHistogramName, dispatchCount)
+	span.SetAttributes(attribute.Float64(dispatchCountHistogramName, dispatchCount))
+	dispatchCountHistogram.WithLabelValues(
+		s.serviceName,
+		methodName,
+	).Observe(dispatchCount)
+
+	requestDurationHistogram.WithLabelValues(
+		s.serviceName,
+		methodName,
+		utils.Bucketize(uint(datastoreQueryCount), s.requestDurationByQueryHistogramBuckets),
+		utils.Bucketize(uint(dispatchCount), s.requestDurationByDispatchCountHistogramBuckets),
+	).Observe(float64(time.Since(start).Milliseconds()))
+
+	return &openfgav1.ListUsersResponse{
+		Users:         resp.GetUsers(),
+		ExcludedUsers: resp.GetExcludedUsers(),
+	}, nil
+}
+
+func userFiltersToString(filter []*openfgav1.UserTypeFilter) string {
+	var s strings.Builder
+	for _, f := range filter {
+		s.WriteString(f.GetType())
+		if f.GetRelation() != "" {
+			s.WriteString("#" + f.GetRelation())
+		}
+	}
+	return s.String()
+}
```

Comment:
Similar comment here as in https://github.com/openfga/openfga/pull/1433#discussion_r1610717643

At minimum (wihtout introducing some clever type) these functions should be colocated with one another. Going to/from a string `UserFilter` to a protobuf and vice versa belong together.

Reply:
The difference here is that `userFiltersToString` is used for the actual production codepath and not solely for testing. Although they seem like they should be grouped, I'd prefer to keep these functions residing nearby where they're used. Once we establish a broader pattern for proto conversion functions (which we have several of), we can address. Otherwise, we'll just be revisiting this problem periodically.

Reply:
Improvements to be made to helpers on our protos :). Doesn't need to block this PR.